lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,283 @@
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return CodeSnippetRenderer.renderSnippetLine(dom, tmpl, {
+      number: ' ',
+      content: highlight.message,
+      contentType: LineContentType.MESSAGE,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineVisibility} visibility
+   * @return {Element}
+   */
+  static renderOmittedLines(dom, tmpl, visibility) {
+    return CodeSnippetRenderer.renderSnippetLine(dom, tmpl, {
+      number: '…',
+      content: '',
+      visibility,
+      contentType: LineContentType.EMPTY,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-code-snippet__content', tmpl);
+    const snippet = dom.find('.lh-code-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
+    nonLineSpecificHighlights.forEach(h =>
+      snippet.append(CodeSnippetRenderer.renderHighlightMessage(dom, tmpl, h))
+    );
+
+    snippet.append(CodeSnippetRenderer.renderSnippetLines(dom, tmpl, details));
+
+    // If expanded view still doesn't include all lines then indicate that that
+    const firstLineIsVisible = lines[0].number === 1;
+    const lastLineIsVisible = lines[lines.length - 1].number === lineCount;
+    if (!firstLineIsVisible) {
+      snippet.prepend(CodeSnippetRenderer.renderOmittedLines(dom, tmpl,
+         LineVisibility.WHEN_EXPANDED
+      ));
+    }
+    if (!lastLineIsVisible) {
+      snippet.append(CodeSnippetRenderer.renderOmittedLines(dom, tmpl,
+        LineVisibility.WHEN_EXPANDED
+      ));
+    }
+
+    return template;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @returns {DocumentFragment}
+   */
+  static renderSnippetLines(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+    const collapsedLines = Util.filterRelevantLines(lines, highlights, 2);
+    const hasLineSpecificHighlights = highlights.some(h => typeof h.lineNumber === 'number');
+    const hasOnlyNonLineSpecficHighlights = highlights.length > 0 && !hasLineSpecificHighlights;
+
+    const lineContainer = dom.createFragment();
+
+    // Keep track of whether a highlighted line has been seen, because we don't want this code
+    // to render an omitted lines indication before the first line
+    let hasSeenHighlight = false;
+    for (let lineNumber = 1; lineNumber <= lineCount; lineNumber++) {
+      const {line, previousLine} = getLineAndPreviousLine(lines, lineNumber);
+      const {line: collapsedLine, previousLine: collapsedPreviousLine} =
```

Comment:
WDYT about some comments or more descriptive names here?

is this right?

- `!previousLine` is just telling us if this current line is the start of a visible block and we need to render the omitted lines placeholder for the stuff that came before?
- `collapsedLine` is just telling us if this current line should be shown even when the snippet is collapsed?
- `!collapsedPreviousLine` is just telling us if this current line is the start of a visible block and we need to render the omitted lines placeholder for the stuff that came before?

Reply:
I've cleaned up the rendering logic a bit more, though it's still not exactly nice code 🤔